Remove non-determinism in Llvm parser and Internal representation

xeren opened the pull request:
Removes two sources of non-determinism of the model checker:
- Identity-based iteration order of phi nodes in the LLVM parser
- Identity-based iteration order of relations in the memory model

hernanponcedeleon reviewed (commented):
I understand that the `LinkedHashSet` guarantees determinism for the iteration order of relations, but I still do not see which part of the change guarantees iteration order of phi nodes.

I assume this is change that does it
```
for (final Map.Entry<BlockPair, List<Event>> phiNode : phiNodes.entrySet()) // before
for (final Map.Entry<Label, List<Event>> phiNode : phiNodes.getOrDefault(block.label, Map.of()).entrySet()) // after
```
but I do not really understand why.

It seems the filters in the Wmm class where not used at all, so it makes sense to get rid of them. However, does anybody remember what the role of those was? While I see the role of filters in programs, I have no clue what a filter might be in a memory model.

xeren commented:
>> It seems the filters in the Wmm class where not used at all, so it makes sense to get rid of them. However, does anybody remember what the role of those was? While I see the role of filters in programs, I have no clue what a filter might be in a memory model.
>
> I think the usage of filters was removed by Renè's PR that added unary relations. He probably forgot to remove the map in the WMM class.

Exactly.

Besides `Event` and `Relation`, there is another class in the IR that probably should become value-based: `Function`.  By extension, `Thread`, `Register` and `Expression` are currently identity-based.
`Function` does not have a persistent identifier.  There already exist two candidate fields.  Each of them has to be altered slightly.

- `Function.id` is set on creation and modified multiple times during program processing.  It also has multiple purposes:
  - Usually corresponds to the index in `program.functions` resp. `program.threads`.
  - Identifies the function for log messages and visualizations.
  - Partially identifies SMT variables for registers.
  - Is the numeric value of the thread handle used by the `pthread_*` intrinsics and the HSA and LOBE progress models.
  - Used in `"END_OF_T"` and `"RETURN_OF_T"` labels.

- `Function.name` is set on creation and modified once in `SymmetryReduction`.  Its purposes currently are:
  - Corresponds to the identifier in the source file, if parsed.
  - Sometimes used in log messages and visualizations, especially with `SyntacticContext`.
  - Used in `Program.getFunctionByName(String)`.
  - Metadata: Name of Init threads starts with `"Init_"`, Litmus threads start with "P", Threads spawned by `pthread_create` share the name with the function argument.

I would prefer `name` to be unique in the program and the sole solution for most of the above situations.  I would move `id` into `Thread` and reduce its usages to intrinsics and progress models.

xeren commented on dartagnan/src/main/java/com/dat3m/dartagnan/expression/type/VoidType.java:
```diff
@@ -10,4 +10,9 @@ public final class VoidType implements Type {
+    @Override
+    public int hashCode() {
+        return 1;
+    }
```
Afaik the canonicalizer in `TypeFactory` might be the only instance.

hernanponcedeleon commented:
I am working on adding native compilation and in there, the non-determinism this PR tries to fix is more visible. On SVCOMP benchmarks I get a bunch of ERROR due to a `The jump does not belong to a loop` exception from loop unrolling when dumping the bounds to a file.

It would be great to finish this PR soon so I can make an SVCOMP run with a larger timeout and create a PR for the native compilation.
